### ISSUE TITLE
Massive async shard fetch requests cost lots of heap memories on master node.

### ISSUE THREAD

**Issue by howardhuanghua** [elastic/elasticsearch#80694]
**Elasticsearch version** : 7.10

**JVM version** (`java -version`): JDK 11

**Description of the problem including expected versus actual behavior**:
In https://github.com/elastic/elasticsearch/pull/77991 we solved asnyc shard fetch resposes memory consumption issue.
But we found async shard fetch reqeusts also cost lots of heap memories. Here is our production env for this exception case:
Data nodes number: 75
Dedicate master nodes number: 3
Master node resource: 2 Core cpus, 8GB physical memory, 4GB heap memory.
Total shards number: 15000

When the new master has been elected after full cluster restart, the elected master heap memory would be used up for several seconds. We dump the memory and found netty inflight sending request used lots of heap:
![image](https://user-images.githubusercontent.com/15325213/141690475-1ab0b0db-0a4f-440b-90a1-d65568460cee.png)

Each `WriteOperation` should be single shard request to specific node (16k buffer size per each):
![企业微信截图_16367325062007](https://user-images.githubusercontent.com/15325213/141664167-3fb1b728-758f-48f1-aa67-607d8940b39c.png)

From `Netty4MessageChannelHandler` class we could see a `queuedWrites`, messages are flushed asynchronously:
https://github.com/elastic/elasticsearch/blob/025dbdc73e228c71025ae599bec2bbf12e1720fe/modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java#L42

So besides cutting fetch shard response, we also need to handle massive shard sending requests.

**Comment by DaveCTurner** [comment]
Relates https://github.com/elastic/elasticsearch/issues/77466

**Comment by DaveCTurner** [comment]
I can see that this might be a problem: 15k shards × 75 nodes × 16kiB buffer for each message ≈ 17GiB of memory churn when the cluster state recovers.

**Comment by howardhuanghua** [comment]
That's it. Finally we scale up heap memory to 16GB and cluster got recovered.

**Comment by howardhuanghua** [comment]
The main problem is that we fetch each shard from all the data nodes, actually only specific single data node contains the right shard copy. Meanwhile, we don't need fetch all the shards at the same in the begining , as we have concurrent allocation throttle. If we restart single data node, we could capture the unassigned shard target node info in disassociating left node event. Then during shard recovery, we could fetch shard from single target node, or the nodes contain this shard's primary + replicas. This is what we have already done in our local test env. But if in full cluster restart case, we haven't stored any shard routing info during cluster state persistent, we don't know which specific node the unassigned shard that was allocated previously. We could discuss more details, it’s our pleasure to optimize this part if possible.

**Comment by DaveCTurner** [comment]
I suspect the problem is primary allocation (i.e. `internal:gateway/local/started_shards`) rather than replica allocation (i.e. `internal:cluster/nodes/indices/shard/store`). Replica allocation messages are naturally throttled because they happen as the primaries come online.

I have a couple of ideas for addressing this:

1. We could batch the requests up so that there's only a limited number of requests in flight at once. Maybe just one-at-a-time. Rather than sending each request right away if there's already one in flight then we could just add them to the next batch to be sent.

2. We could keep track of which shards are definitely not on each node and skip sending requests to nodes that don't have anything to offer.

I prefer the first idea I think. The second is effectively a cache which we could populate at cluster startup fairly easily but it would be tricky to implement the refresh or invalidation logic correctly. Batching has the disadvantage that a shard on a slow/broken disk will hold up replies about other shards but I don't think that's a big deal in practice since a slow/broken disk on a node causes other issues.

**Comment by DaveCTurner** [comment]
I won't have time to work on this myself in the near future, but if you would like to work on this then please do.

**Comment by howardhuanghua** [comment]
@DaveCTurner Thanks for the suggestion. Just to confirm details about batching mode, one of the idea is that group all the unassigned primary shards together, send to each node, that means each node only has single fetch request. Otherwise we still need to prepare single request per shard for each node?

**Comment by DaveCTurner** [comment]
Yes that's about right. It won't be _all_ the unassigned shards, ideally we'd keep the same overall behaviour as today, just using fewer transport messages.

Having looked at this a bit more I can see value in adding batching to the replica allocator too. There are cases (e.g. a big network partition) where we'd send a large number of `internal:cluster/nodes/indices/shard/store` messages today.

**Comment by howardhuanghua** [comment]
Hi @DaveCTurner , I quick implemented a draft to batch the async shard fetching request for primary first.
Please help to check it's the right direction, the main logic is in `InternalPrimaryShardAllocator`, there are still some works need to be done.
